jf_agent/git/gitlab_v3_client.py

```diff
@@ -0,0 +1,314 @@
+    def list_project_branches(self, project_id):
+        project = self.get_project(project_id)
+        return project.branches()
+
+    def list_project_merge_requests(self, project_id, state_filter=None):
+        project = self.get_project(project_id)
+        mergerequests = project.merge_requests()
+        if len(mergerequests) > 0:
+            if state_filter:
+                mergerequests = [entry for entry in mergerequests if entry.state in state_filter]
+            mergerequests.sort(key=lambda x: x.created_at, reverse=True)
+            mergerequests = self.add_v4_attrs('mergerequest', mergerequests, project)
+            return mergerequests
+        return mergerequests
+
+    def list_project_commits(self, project_id, since_date):
+        project = self.get_project(project_id)
+        commits = self.add_v4_attrs('commit', list(project.commits()), project)
+        if len(commits) > 0:
+            if since_date:
+                commits = [commit for commit in commits if commit.created_at > since_date]
+                commits.sort(key=lambda x: x.created_at, reverse=True)
+            return commits
+        return []
+
+    def get_project_commit(self, project_id, sha):
+        project = self.get_project(project_id)
+        mrg_commit = project.find_commit(id=sha)  # only searches master branch by default
+        if mrg_commit:
+            try:
+                commit = self.add_v4_attrs('commit', [mrg_commit], project)
+                return commit[0]
+            except gitlab3.exceptions.GitLabException:
+                return None
+        return None
+
+    def get_branch_commit(self, branch_name, sha):
+        branch = self.get_branch(branch_name)
+        branch_commit_id = branch.commit['parent_ids'][0]
+        if sha == branch_commit_id:
+            project = branch._parent
+            parent_commit = self.get_project_commit(project.id, branch_commit_id)
+            merge_commit = project.Commit(parent=parent_commit)
+            # Add more v4 attrs here...
+            commit = self.add_v4_attrs('commit', [merge_commit], project)
+            return commit[0]
+        else:
+            return None
+
+    def get_mergerequest_commits(self, mergerequest, project=None):
+        if project == None:
+            project = self.client.find_project(id=mergerequest.project_id)
+        commits = mergerequest.get_commits()
+        commits.sort(key=lambda x: x['created_at'], reverse=False)
+        payload = []
+        if commits:
+            # search project for merged PR commits
+            for i, commit_dict in enumerate(commits):
+                commit = self.get_project_commit(project.id, commit_dict['id'])
+                if commit:
+                    payload.append(commit)
+
+            # manually create commit object if api returns no PR commits have been merged
+            if payload:
+                return payload
+            else:
+                all_requests = project.merge_requests()
```

Comment:
Hmmm I think I need to understand this block better... maybe you can walk me through this on Monday...

Reply:
Agreed, this is the result of many iterations of tests. I have a donut mtg from 4-4:30. I'll ping you after that to review :)